ENH: special: add the Wright Omega function

person142 opened the pull request:
This adds @pierslawrence's implementation of the Wright Omega function to `special` as discussed in #2851. Since we don't support C99 complex arithmetic the original C code is ported to C++. The tests generally look good, though there are what appear to be some overflow issues for large arguments. They might be caused by this line:

https://github.com/person142/scipy/blob/48f02f1ed7a2af6041fc4eb45726f23ab73faa41/scipy/special/wright.cc#L259

(In particular the division by `z*z*z*`; the product could overflow.) Or perhaps I introduced an error while porting; haven't investigated much since at this point it seems best to leave the code as close to the original as possible.

I am unsure as to whether we should also add the unwinding number as a special function since it's a one-liner, but do not have very strong opinions either way.

person142 commented:
Ok, the build is failing because Travis is using gcc 4.6 which only includes experimental support for the C++11 `cfenv`. Since `fenv.h` is C99, is there a portable way to get functions like `fesetround` besides adding our own implementations? (NumPy doesn't have them AFAICT?)

person142 commented:
Can probably use FreeBSD's implementation:

https://github.com/freebsd/freebsd/blob/af3e10e5a78d3af8cef6088748978c6c612757f0/lib/msun/mips/fenv.h#L127

though it seems to imply that on ARM changing the rounding mode won't work anyway.

person142 commented:
Whew, finally a happy Travis. Use of `fenv` is eliminated: the code was using it to change the rounding mode before an addition so standalone functions that add with a desired rounding mode are added in `special/_round.h`. They rely on the fact that `error(a + b) = a + b - fl(a + b)` is exactly representable (and computable) in double precision; the sign of the error tells us whether rounding to nearest rounded up or down.

Here's a gist containing the code I was using to test the addition functions:

https://gist.github.com/person142/2eec2572fed3fea7ec22155f204c0aff

person142 commented:
Tweaked the asymptotic series to avoid overflow and also skipped using the iterative scheme in some regimes where it can overflow/try to take a log after underflow. (In those regimes the asymptotic expansions are accurate enough.) The systematic tests now pass on the entire complex plane even with `nan_ok=False`; added tests near the branch cuts which also look good.

Think I'm done fiddling with things.

pv commented on scipy/special/_c99compat.h:
```diff
@@ -0,0 +1,55 @@
+
+
+int
+sc_signbit(double x)
```
Probably not, because `#include <complex>` includes `<cmath>`.
It's really a PITA that we can't use C99 or C++11 --- the former won't work on MSVC, and the latter apparently requires passing compiler-dependent flags in.

person142 replied:
Yeah I used `npy_signbit` at first but then g++ undef'd it on Travis. :-(

charris replied:
Hmm, what do you mean by g++  undef'd `npy_signbit`?

person142 replied:
I didn't look into it too carefully, but I assumed it was this issue: https://github.com/scipy/scipy/issues/5689#issuecomment-170259179. (The CI build complained about not being able to find `signbit`.)

person142 replied:
So my understanding (maybe wrong) was: the CI machine has C99, so it hits

`#define npy_signbit(x) signbit((x))`

in `npy_math.h`. Then we hit `#undef signbit` in `cmath`. But then there's no C++11, so we end up having no `signbit`. Maybe this could also be fixed by shuffling the includes, though that sounds kind of brittle.

person142 commented:
I didn't need to be using `signbit` since the code didn't need to detect the sign of `0` or `NaN`. Wasn't thinking; it's gone now.

pv commented on scipy/special/_round.h:
```diff
@@ -0,0 +1,64 @@
+
+double add_round_up(double a, double b)
+{
+    double s, e, v;
```
May need to be `volatile double` --- is the compiler allowed to keep stuff in fp extended precision registers, so you don't get the truncation as expected? Does this break with `-O3` etc? How do you test that this works?

pv replied:
https://gcc.gnu.org/wiki/FloatingPointMath --- gcc claims to do it as expected on SSE2, but not necessarily with x87 fpu
https://randomascii.wordpress.com/2012/03/21/intermediate-floating-point-precision/

person142 replied:
This would very much break with `-O3` since (IIRC) `-O3` enables unsafe floating point optimizations. Tests I was using to see if it works were mentioned here: https://github.com/scipy/scipy/pull/6653#issuecomment-251879577; would be possible to put the tests in SciPy itself if we only run them when we detect that `fenv` is present. Shall I?

pv replied:
It may also break on without unsafe optimizations on x87, and I'm not sure if C (or C99) standards specify the precision used in intermediate results.

pv replied:
If you need just controlled truncation, I think assignment to a `volatile double` variables is OK.

pv commented on scipy/special/wright.cc:
```diff
@@ -203,6 +206,18 @@ wright::wrightomega_ext(complex<double> z, complex<double> *w,
+	  sf_error("wrightomega", SF_ERROR_UNDERFLOW, "underflow in exponential series");
+	  /* Skip the iterative scheme because it computes log(*w) */
+	  e = NULL;
+	  r = NULL;
```
I think these assignements don't do anything?

pv commented on scipy/special/wright.cc:
```diff
@@ -67,9 +67,11 @@
 /*                                                                    */
 /*   r  --  double complex*                                           */
 /*          Pointer to penultimate residual r_k = z - w_k - log(w_k)  */
+/*          NULL if the iterative scheme is not used.                 */
```
Cannot return NULL since it's not a pointer-to-pointer?

pv commented on scipy/special/wright.cc:
```diff
@@ -222,7 +237,24 @@ wright::wrightomega_ext(complex<double> z, complex<double> *w,
+	/* Series is accurate and the iterative scheme could overflow */
+	{
+	  e = NULL;
+	  r = NULL;
```
Unnecessary assignments.

pv commented on scipy/special/wright.cc:
```diff
@@ -0,0 +1,389 @@
+	/* Series is accurate and the iterative scheme could overflow */
+	{
+	  e = NULL;
+	  r = NULL;
```
As above.

pv commented on scipy/special/wright.cc:
```diff
@@ -0,0 +1,389 @@
+	/* Series is accurate and the iterative scheme could overflow */
+	{
+	  e = NULL;
+	  r = NULL;
```
As above.

pv commented on scipy/special/wright.cc:
```diff
@@ -0,0 +1,389 @@
+    {
+      pz=log(z);
+      *w = z - pz;
+      fac = pz/z;
```
I'm assuming these branches were tested.

person142 replied:
They should be covered in the `TestSystematic` test--if you go back to https://github.com/scipy/scipy/pull/6653/commits/347d7d384585ad8f4046455bdb80faa31bc0d73e but flip the `nan_ok` flag in `TestSystematic.test_wrightomega` to `False` you'll get around 1,200 failures, which these changes are trying to address.

person142 commented:
Added tests to SciPy for `add_round_up` and `add_round_down` (and marked everything as volatile); they pass locally and fail on Travis so the answer to "is it portable" seems to be "definitely not". Back to square 1...

pv commented:
Maybe you can try removing the `assert_` from the test and add more precision `{:.21g}`, so that it will show which points failed. I.e., is it even clear fesetround/fegetround work on the travis gcc?

person142 commented on scipy/special/_test_round.pyx:
```diff
@@ -0,0 +1,47 @@
+                       .format(sample2[i], sample2[i], std, res))
+    if nfail:
+        s = "{}/{} failures with mode {}.".format(nfail, size, mode)
+        msg = s + msg
```
Urg.

person142 commented:
Indeed, might be an issue with `fenv` on the Travis machines. For example one failure is:

```
a = -5.41562869078764134668e+306
b = -9.46574644932701959158e+306
a + b in round up mode = -1.48813751401146621857e+307
a + b with add_round_up = -1.48813751401146596909e+307
```
But locally I get:
```
a + b in round up mode = -1.48813751401146596909e+307
```
Also tested on Ubuntu 16.04 (gcc 5.3.1, glibc 2.23) and tests passed there. (Other local tests were on OS X 10.11.6.)

person142 commented:
Now:

- Travis build is succeeding because the `add_round_*` tests are being skipped--I modified `have_fenv` (which controls whether the tests run) to also check whether `fesetround` is working (before just tested `fegetround`), and apparently it is not. (i.e. it returns an error status when you try to change the rounding mode.)
- After fixing a bug I introduced when modifying the asymptotic series the relative error on the tests is as good as the original (and better in the asymptotic regime where there were overflows). Those interested in checking this can look at: https://github.com/person142/scipy/tree/wrightomega-original-code. (The only change in this branch is the port to C++; i.e. it's still using `fenv` and there are no modifications to the asymptotic regime.)

pv commented:
I get deprecation warnings from test_round:
```
/home/pauli/prj/scipy/scipy/scipy/special/tests/test_round.py:12: DeprecationWarning: This function is deprecated. Please call randint(-9223372036854775808, 9223372036854775807 + 1) instead
  _test_round.test_add_round(10**5, 'up')
/home/pauli/prj/scipy/scipy/scipy/special/tests/test_round.py:12: DeprecationWarning: This function is deprecated. Please call randint(-9223372036854775808, 9223372036854775807 + 1) instead
  _test_round.test_add_round(10**5, 'up')
./home/pauli/prj/scipy/scipy/scipy/special/tests/test_round.py:18: DeprecationWarning: This function is deprecated. Please call randint(-9223372036854775808, 9223372036854775807 + 1) instead
  _test_round.test_add_round(10**5, 'down')
/home/pauli/prj/scipy/scipy/scipy/special/tests/test_round.py:18: DeprecationWarning: This function is deprecated. Please call randint(-9223372036854775808, 9223372036854775807 + 1) instead
  _test_round.test_add_round(10**5, 'down')
```
Otherwise LGTM.

The corner-case branches before `/* Choose approximation based on region */` in wright.cc are not tested (`python runtests.py -g -s special -m full --gcov; python runtests.py --lcov-html; firefox build/lcov/index.html`), but probably doesn't block merge.

person142 commented:
Hopefully this fixes the above issues. Indeed, the corner cases were not handling `-0.0` correctly. (Warning: did not run `gcov` myself; llvm builds appear to no longer be compatible and the gcc build keeps failing.)

pv commented:
Thanks, innit goes

pierslawrence commented:
Hello,

I am really appreciative for all of the effort by everyone to integrate this function into scipy. I do have one question about the usual practice regarding real/complex functions within this module.

For real valued z the wrightomega(z) is also real (the same can be said for lambertw(z,k=0) on (1/e,inf) and lambertw(z,k=-1) on (-1/e,0)). 

What is the normal practice for dealing with these cases? Should we expect to get back a real value, or is it usual to return a complex type with Im(w)=0.?

Kind regards,

Piers

person142 commented:
> Should we expect to get back a real value, or is it usual to return a complex type with Im(w)=0.?

Ideally get back a real value, but we haven't always been very consistent about that. See e.g. https://github.com/scipy/scipy/pull/7983 where we fixed this issue for `loggamma`.

pierslawrence commented:
Sorry it took me so long to follow up on this real extension to the Wright omega function.

As mentioned in the first post of this thread, we really do need to avoid the overflow introduced by the z*z*z term in the series about infinity. As it turns out, we actually only need the first two terms (x-log(x)+log(x)/x) in the series about infinity on the real axis, and we can avoid the overflow problem. The same could probably be done for the complex code, but it would need testing for the accuracy.

Similarly, we need only the first term (exp(x)) of the series about negative infinity, and this is good enough on (-inf,-2] with two iterations of the Fritsch, Schafer, Crowley (FSC) iteration.

For the remaining region (-2,1] we only need a low accuracy approximation, and I suggest exp(2/3*(x-1)) which is continuous with respect to the other series (not the derivatives, obviously) and is good enough to obtain double precision accuracy in two iterations of FSC.

I attach my implementation of the real Wright omega function, if someone with more knowledge of integrating into scipy can work this in, I would appreciate it (especially since the complex implementation seemed to have many issues with the build system).

Regards,

Piers Lawrence

[wright_real_dispatch.txt](https://github.com/scipy/scipy/files/3467835/wright_real_dispatch.txt)

person142 commented:
Hey @pierslawrence, could you open a new issue for the real implementation? It will have a lot more visibility that way and could be a good first project for someone looking to get familiar with the special internals.
